beijbom/coralnet
# Spacer integration

This addresses most of https://github.com/beijbom/coralnet/issues/264.

# Summary
I wanted to try not making this PR too large. But I'm afraid I failed. Here are the main changes:
- Use the new spacer in `vision_backend/tasks.py` and `task_helpers.py`.
- Use a swappable queue to allow `SQSQueue` for pushing jobs to AWS and `LocalQueue` for processing jobs locally. The latter is for unit-tests and reg-tests. Note that the only thing we change here is the Queue. The actual processing is done by `pyspacer` in both cases.
- PEP8 compliance in the files I touched. I didn't go over all files in the app.
- Refreshed backend regression tests.
- Deleted `vision_backend/scrips.py`.

# How to test
All unit-test should pass. Run regression tests by `python manage.py vb_regtests`. There are several ways to run the tests: check out `python manage.py vb_regtests --help` for details.
## Local eager
This is the simplest setting. Local backend, no celery needed, local queue.
```
CELERY_ALWAYS_EAGER = True
SPACER_QUEUE_CHOICE = 'vision_backend.queues.LocalQueue'
from .storage_local import *
```

## Local celery
Here everything is still local, but jobs are executed asynchronously using celery. Make sure the celery worker is running before starting the test: `celery -A config worker`.
```
CELERY_ALWAYS_EAGER = False
SPACER_QUEUE_CHOICE = 'vision_backend.queues.LocalQueue'
from .storage_local import *
```

## S3 celery
Still using the LocalQueue, but storage is now on S3. 
```
CELERY_ALWAYS_EAGER = False
SPACER_QUEUE_CHOICE = 'vision_backend.queues.LocalQueue'
from .storage_s3 import *.
```

## AWS ECS cluster. 
This uses the AWS SQS and ECS to process the jobs. The jobs are submitted to a test_queue, so this is safe to run.
```
CELERY_ALWAYS_EAGER = False
SPACER_QUEUE_CHOICE = 'vision_backend.queues.SQSQueue'
from .storage_s3 import *
```
Etc. etc. Most combinations work. The only requirement is that if you use the `SQSQueue` you need to use the `storage_s3` setting. Also, if you use `.storage_local` you need to add these lines to your personal dev_settings to allow access to fixtures and spacer models from s3.
```
AWS_ACCESS_KEY_ID = get_secret('AWS_ACCESS_KEY_ID')
AWS_SECRET_ACCESS_KEY = get_secret('AWS_SECRET_ACCESS_KEY')
os.environ['SPACER_AWS_ACCESS_KEY_ID'] = AWS_ACCESS_KEY_ID
os.environ['SPACER_AWS_SECRET_ACCESS_KEY'] = AWS_SECRET_ACCESS_KEY
```

# Notes
- I removed `FORCE_NO_BACKEND_SUBMIT` since now we can always submit jobs now, we just need to make sure its the right queue.
- I added a `FORCE_DUMMY_EXTRACTOR` to be set for tests. This forces tasks to use the spacer `dummy` extractor, which is very fast.
- I added a new secret: `SPACER_JOB_HASH`. This is a unique string that ensures we only collect jobs for the correct server.
- I added configs for the two sets of SQS I created.
- I had to mute the logger during testing. I think it has to do with spacer using the logging module.
- I tried enabling `SuccessTest.test_done()` in `vision_backend_api/tests/test_deploy.py`. To do that I needed to actually train a classifier, which we now can do quite quickly. So I used a similar design as in `TrainClassifierTest` in vision_backend/tests/test_tasks.py. But for some reason, it didn't work here. I suspect is has to do with the test base classes. 
- The code in `regtests_utils.py` is horrible but I couldn't be bothered fixing it. I wasn't quite sure how and didn't feel like going down that hole.

- [x] Merged in the swappable backend from https://github.com/beijbom/coralnet/pull/265.
- [x] Port extract_features.
- [x] Switch to local spacer dev-version.
- [x] Port train_classifier.
- [x] Write job_token parser, and fix previous classifier updates.
- [x] Port classify_features.
- [x] Port classify_image (aka deploy).
- [x] Update deploy parse test.
- [x] Make all unit-tests pass.
- [x] Fix `add_scores` and `add_annotations`.
- [x] Add spacer_job_hash.
- [x] Rename `MockBackend` to `MockQueue`, etc.
- [x] Remove FORCE_NO_BACKEND_SUBMIT?
- [x] Change from print to log in spacer.
- [x] Make sure new `valres` is parsed correctly.
- [x] Run reg-tests using MockQueue with both storage classes.
- [x] Write ValRes size tests to catch previous bug.
- [x] Make backwards compatibility tests for ValRes.
- [x] Make new spacer release.
- [x] Switch to new spacer release.
- [x] Update reg-tests management command.
- [x] Clean up scripts.py in vision_backend.
- [x] Update spacer ECS with latest docker.
- [x] Run reg-tests using SQSQueue with S3 storage.


This PR has two minor "friend-PRs" on the spacer side. These are already merged to master and included in the `0.2.6` release.
https://github.com/beijbom/pyspacer/pull/25
https://github.com/beijbom/pyspacer/pull/26
